records/2021-09-27.batch-celery-save.md

```diff
@@ -0,0 +1,116 @@
+individually in one transaction, except you have potentially thousands of items
+remaining afterward. Ideally, we want to buy/save all for all of these in one
+go.
+
+### Current notification flow through Celery tasks
+
+> **INFO**: [Celery](https://docs.celeryproject.org/en/stable/getting-started/introduction.html) is a Python library to process a message through a workflow of tasks which offers reliability mechanism (e.g. retry), monitoring, concurrency (to execute multiple work at the same time), automatic scheduling of tasks and allows for ways to scale up the processing.
+
+Our current workflow of Celery tasks for processing an email or SMS at the moment
+is simplified with the next diagram.
+
+![Current workflow of a notification save](https://raw.githubusercontent.com/cds-snc/notification-adr/main/records/diagrams/2021-09-27.batch-celery-save/current-workflow.jpg?raw=true "Current workflow of a notification save")
+
+## Options
+
+Listing of potential solutions and necessary architectural changes will be listed in
+this section. Feel free to list your own idea in here via a proposed change to this
+document.
+
+### [celery-batches](https://github.com/clokep/celery-batches)
+
+Around the development of [Celery version 3](https://docs.celeryproject.org/en/3.1/reference/celery.contrib.batches.html), there was an option to batch the content
+of a queue and have a task flush it all when one of the two constraint are met:
+the number of maximum requests or a timeout. For example, with a limit of 1000
+requests maximum and a timeout of 5 seconds, if 1,000 requets would be filled in 2
+seconds, then the task would trigger and flush all of the 1,000 requests buffered
+in the current task. Alternatively, if only 230 requests filled up and the 5
+seconds timeout completed, then all of the 230 requests would get flushed to
+the task.
+
+This is a transparent and easy mechanism that could get leveraged in Celery in order
+to achieve the batch save. Unfortunately, that feature was always in an experimental
+state in Celery version 3 and disappeared in subsequent versions. An individual
+contributer on GitHub [made a fork of the experimental feature](https://github.com/clokep/celery-batches) and has adapted the latter for recent
+versions of Celery. This looked promising unfortunately there are two issues by
+using this library:
+
+1. It still can be considered in an experimental stage as this does not support common basic features offered by Celery such [as a retry mechanism](https://github.com/clokep/celery-batches/issues/10).
+2. The integration is not that easy with Flask (although probably not impossible) as described by the Flask integration recipe [in their documentation](https://flask.palletsprojects.com/en/2.0.x/appcontext/). [An issue was opened](https://github.com/clokep/celery-batches/issues/35) for curiosity sake to see if that could be done but even then, the issue described in #1 is a blocker.
+
```

Comment:
If this approach is not possible in the foreseeable future, should we even have it here as an option?

Reply:
If someone thinks they can solve it with `celery-batches` in the future and don't see it listed in the options over here, they might re-do the investigative work. 😅